### PR TITLE
Issue98: give topic suggestions to the instructor

### LINKED ISSUE #98: Give topic suggestions to the instructor
When creating a question, the instructor should be able to see a list of previously written topics to select from

### PR DESCRIPTION
**Pull request by esteeroramy**
fixes #98 
image

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -810,6 +810,32 @@ var submitQuestionRating = function (req, res) {
...
+    }
+
+    questions.getQuestionsList(function(err, docs){
+        if (err) {
+            return res.status(500).send('could not get the list of questions topics');
+        }
+
+        var topicsList = [];
+        for (var i in docs) {
+            if (topicsList.indexOf(docs[i].topic) === -1) {
```
Comment (mehtadarshan1) [inline server.js:796]: Not very efficient on server side, Takes a bit of a hit when you have to look through all questions and compile a list each and overtime. Should be stored dynamically in DB

File: public/js/admin.js [hunk 2]
```diff
@@ -157,6 +158,21 @@ var displayQuestionForm = function() {
                 submitQuestionForm();
             });
             $('select').material_select();
+
+            // gets the updated topics list
+            getQuestionsTopicsList();
```
Comment (mehtadarshan1) [inline public/js/admin.js:163]: Same code in editQuestion as well, please simplify

File: public/js/admin.js [hunk 3]
```diff
@@ -594,6 +610,21 @@ var editQuestion = function(qid) {
                 submitQEditForm(qid);
             });
             setRating(data.qrating);
+
+            // gets the updated topics list
+            getQuestionsTopicsList();
+            autocompleteTopics = {};
```
Comment (mehtadarshan1) [inline public/js/admin.js]: duplicate code

File: server.js [hunk 4]
```diff
@@ -810,6 +810,32 @@ var submitQuestionRating = function (req, res) {
...
 }
 
+// questions list of topics
+app.get('/questionsListofTopics', function(req, res){
+    if (!req.session.user) {
+        return res.redirect('/');
+    }
+
+    if (req.session.user.type !== common.userTypes.ADMIN) {
+        return res.status(403).send('Permission Denied');
```
Comment (mehtadarshan1) [inline server.js:786]: returning a 403, however it is not being caught in the ajax call

File: public/js/admin.js [hunk 5]
```diff
@@ -736,3 +767,21 @@ var toggleButtonVisibility = function(){
...
+var getQuestionsTopicsList = function () {
+    $.ajax({
+        async: false,
+        type: 'GET',
+        url: '/questionsListofTopics',
+        success: function(data) {
+            questionTopicsList = data;
+        },
+        error: function(data){
+            if (data['status'] === 401) {
```
Comment (mehtadarshan1) [inline public/js/admin.js]: 401 status never gets returned

**Review by mehtadarshan1 — CHANGES_REQUESTED** [review]

**Review by Mohammed-JS-Khan — APPROVED** [review]
Nice. Just have to address @mehtadarshan1 's points. I agree with them.

File: public/js/admin.js (hunk 5 above)
Comment (esteeroramy) [reply public/js/admin.js]: Ahmed:`401` is the session checked and it's always returned in every request we do

**Comment by esteeroramy** [comment]
Addressed comments @mehtadarshan1

**Review by Mohamed-A-Mohamed — APPROVED** [review]
Nice work

File: public/js/admin.js (hunk 5 above)
Comment (mehtadarshan1) [reply public/js/admin.js]: okay

File: server.js (hunk 4 above)
Comment (mehtadarshan1) [reply server.js:786]: are you not catching this in the ajax?

File: server.js (hunk 1 above)
Comment (mehtadarshan1) [reply server.js:796]: didn't address this

**Review by mehtadarshan1 — DISMISSED** [review]

File: server.js (hunk 1 above)
Comment (ahmedqarmout2) [reply server.js:796]: I did some scale testing and this is not an issue

Comment (esteeroramy) [reply server.js:796]: thank you @ahmedqarmout2

**Review by ahmedqarmout2 — APPROVED** [review]

**Pull request merged**